consensus/istanbul/backend/internal/enodes/val_enode_db.go

```diff
@@ -212,78 +213,103 @@ func (vet *ValidatorEnodeDB) String() string {
 }
 
 // GetEnodeURLFromAddress will return the enodeURL for an address if it's known
-func (vet *ValidatorEnodeDB) GetEnodeURLFromAddress(address common.Address) (string, error) {
+func (vet *ValidatorEnodeDB) GetNodeFromAddress(address common.Address) (*enode.Node, error) {
 	vet.lock.RLock()
 	defer vet.lock.RUnlock()
-	entry, err := vet.getAddressEntry(address)
+	entry, err := vet.GetAddressEntry(address)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
-	return entry.enodeURL, nil
+	return entry.Node, nil
 }
 
-// GetAddressFromEnodeURL will return the address for an enodeURL if it's known
-func (vet *ValidatorEnodeDB) GetAddressFromEnodeURL(enodeURL string) (common.Address, error) {
+// GetAddressFromNodeID will return the address for an nodeID if it's known
+func (vet *ValidatorEnodeDB) GetAddressFromNodeID(nodeID enode.ID) (common.Address, error) {
 	vet.lock.RLock()
 	defer vet.lock.RUnlock()
-	return vet.getAddressFromEnodeURL(enodeURL)
+	return vet.getAddressFromNodeID(nodeID)
 }
 
-func (vet *ValidatorEnodeDB) getAddressFromEnodeURL(enodeURL string) (common.Address, error) {
-	rawEntry, err := vet.db.Get(enodeURLKey(enodeURL), nil)
+func (vet *ValidatorEnodeDB) getAddressFromNodeID(nodeID enode.ID) (common.Address, error) {
+	rawEntry, err := vet.db.Get(nodeIDKey(nodeID), nil)
 	if err != nil {
 		return common.ZeroAddress, err
 	}
 	return common.BytesToAddress(rawEntry), nil
 }
 
+func (vet *ValidatorEnodeDB) GetAllValEnodes() (map[common.Address]*AddressEntry, error) {
+	vet.lock.RLock()
+	defer vet.lock.RUnlock()
+	var entries = make(map[common.Address]*AddressEntry)
+
+	err := vet.iterateOverAddressEntries(func(address common.Address, entry *AddressEntry) error {
+		entries[address] = entry
+		return nil
+	})
+
+	if err != nil {
+		vet.logger.Error("ValidatorEnodeDB.GetAllAddressEntries error", "err", err)
+		return nil, err
+	}
+
+	return entries, nil
+}
+
 // Upsert will update or insert a validator enode entry; given that the existing entry
 // is older (determined by view parameter) that the new one
-func (vet *ValidatorEnodeDB) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) error {
+func (vet *ValidatorEnodeDB) Upsert(valEnodeEntries map[common.Address]*AddressEntry) error {
```

Comment:
Discussed offline:

1) Check View to see if there's a need to actually make the update. 
2) Instead of returning an error on old update, return a `UpsertsResults` with inserts#, updates#, etc